Question traverse process double-unescapes HTML content

### Details

If an element code produces a result that includes a string like `&amp;gt;`, the question traverse will convert this into `&gt;`, which will then be interpreted by the browser as just `>`. As far as I could identify, this is a result of this code:

https://github.com/PrairieLearn/PrairieLearn/blob/c4658495b9912ce56ec77881104e42017f12b1c6/apps/prairielearn/python/prairielearn/internal/traverse.py#L66-L76

The problem here is that the text is already unescaped (this function receives `&gt;`), so the call to `html_unescape` will incorrectly "unescape" it twice.

I don't know what the best solution here is. I added a test to traverse_test.py with the following content, and it failed:

```python
def test_traverse_and_replace_ampersand_followed_by_entity_name() -> None:
    def replace(e: lxml.html.HtmlElement) -> ElementReplacement:
        if e.tag == "pl-string":
            return "<div>&amp;lt;div&amp;gt;</div>"
        return e

    html = traverse_and_replace("<pl-string></pl-string>", replace)
    assert html == "<div>&amp;lt;div&amp;gt;</div>"
```

```output
FAILED apps/prairielearn/python/test/traverse_test.py::test_traverse_and_replace_ampersand_followed_by_entity_name - AssertionError: assert '<div>&lt;div&gt;</div>' == '<div>&amp;lt...amp;gt;</div>'
```

--- a/apps/prairielearn/python/prairielearn/internal/traverse.py
+++ b/apps/prairielearn/python/prairielearn/internal/traverse.py
@@ -1,7 +1,6 @@
 from collections import deque
 from collections.abc import Callable, Sequence
 from html import escape as html_escape
-from html import unescape as html_unescape
 from itertools import chain
 
 import lxml.html
@@ -63,17 +62,8 @@ def get_source_definition(
     return f"<{' '.join((element.tag, *attributes))}>"
 
 
-# `lxml` uses `libxml2` under the hood, which does not support the full set
-# of HTML5 named entities:
-# https://gitlab.gnome.org/GNOME/libxml2/-/issues/857
-# This means that with a naive approach, we'd end up double escaping entities
-# like `&langle;` into `&amp;langle;`. To work around this (at least until
-# `libxml2` hopefully adds support for HTML5 named entities), we first
-# unescape the text to get the actual Unicode characters, and then escape them
-# again. Escaping will only escape `&`, `<`, and `>`; it won't escape everything
-# that could possibly be represented by a named entity.
 def prepare_text(text: str) -> str:
-    return html_escape(html_unescape(text))
+    return html_escape(text)
 
 
 def traverse_and_replace(

--- a/apps/prairielearn/python/test/traverse_test.py
+++ b/apps/prairielearn/python/test/traverse_test.py
@@ -333,6 +333,16 @@ def test_traverse_and_replace_angle_brackets() -> None:
     assert html == "<pre><code>&lt;div&gt;</code></pre>"
 
 
+def test_traverse_and_replace_escaped_entities() -> None:
+    def replace(e: lxml.html.HtmlElement) -> ElementReplacement:
+        if e.tag == "pl-string":
+            return "<div>&amp;lt; &amp;langle;v, w&amp;rangle; &amp;gt;</div>"
+        return e
+
+    html = traverse_and_replace("<pl-string></pl-string>", replace)
+    assert html == "<div>&amp;lt; &amp;langle;v, w&amp;rangle; &amp;gt;</div>"
+
+
 def test_traverse_and_replace_trailing_entity() -> None:
     def replace(e: lxml.html.HtmlElement) -> ElementReplacement:
         if e.tag == "div":

